### ISSUE TITLE
SFTP using 2FA with PingID fails without prompting for OTP

### ISSUE THREAD

**Issue by cankaratepe23** [iterate-ch/cyberduck#15076]
**Describe the bug**
Connecting to an SFTP server using Two Factor Authentication with Password + PingID fails with "Gateway authentication failed". Connection works through the terminal with the regular SSH client. This error message is from the destination server and indicates a wrong password to the PingID one time password prompt. A prompt for the PingID password never appears.

Investigating the logs myself, I can see that my initial password is sent and accepted by the server. The PingID prompt is received from the server, is logged, but a prompt for the one time password never pops up.

From what I can understand out of the logs, I see that both the initial user password AND the PingID password are configured as 'keyboard-interactive' challenges. Could it be that Cyberduck is re-using the initial password for the second challenge?

I do not have access (neither read nor write) to the configuration of the server, but the fact that it works over regular SSH, as well as with WinSCP on Windows should indicate the configuration is correct.

**To Reproduce**
1. Attempt connection to an SFTP server using PingID 2FA
2. Enter user password
3. Wait 1 second

**Expected behavior**
After entering the first password, Cyberduck should prompt for the PingID one time password.

**Desktop:**
 - OS: macOS
 - Version Ventura 13.5.1

**Log Files**
[cyberduck.log](https://github.com/iterate-ch/cyberduck/files/12496279/cyberduck.log)

(Note that I have obfuscated the domain names and IP addresses in a way that preserves meaning.)

**Additional context**
Related section of the log file:
```
2023-09-01 11:32:49,453 [Thread-89] DEBUG net.schmizz.concurrent.Promise - Setting <<authenticated>> to `null`
2023-09-01 11:32:49,453 [Thread-89] DEBUG net.schmizz.sshj.userauth.UserAuthImpl - Trying `keyboard-interactive` auth...
2023-09-01 11:32:49,453 [Thread-89] DEBUG net.schmizz.concurrent.Promise - Awaiting <<authenticated>>
2023-09-01 11:32:49,520 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] DEBUG net.schmizz.sshj.userauth.UserAuthImpl - Asking `keyboard-interactive` method to handle USERAUTH_60 packet
2023-09-01 11:32:49,520 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] DEBUG net.schmizz.sshj.userauth.method.AuthKeyboardInteractive - Requesting response for challenge `Gateway password: `; echo=false
2023-09-01 11:32:49,520 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] DEBUG ch.cyberduck.core.sftp.auth.SFTPChallengeResponseAuthentication - Reply to challenge name Gateway authentication and authorization with instruction Please specify the requested information
2023-09-01 11:32:49,951 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] DEBUG net.schmizz.sshj.userauth.UserAuthImpl - Asking `keyboard-interactive` method to handle USERAUTH_60 packet
2023-09-01 11:32:49,951 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] DEBUG net.schmizz.sshj.userauth.method.AuthKeyboardInteractive - Requesting response for challenge `

PingID MFA@mucblbp06.muc.msp.companyco.com
For any questions please advise
https://companyco.com/confluence/display/helpdoc1
https://companyco.com/confluence/display/helpdoc2


Depending on your PRIMARY/default MFA device:
* press Enter for push notification (cell phone)
* type one-time password from your PRIMARY device (desktop app or cell phone)
* press your YubiKey (Yubikey)
* type !select to change MFA method

PingID MFA: `; echo=false
2023-09-01 11:32:49,952 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] DEBUG ch.cyberduck.core.sftp.auth.SFTPChallengeResponseAuthentication - Reply to challenge name Gateway authentication and authorization with instruction Please specify the requested information
2023-09-01 11:32:50,646 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] INFO  net.schmizz.sshj.transport.TransportImpl - Received SSH_MSG_DISCONNECT (reason=BY_APPLICATION, msg=Gateway authentication failed
Authentication failed)
2023-09-01 11:32:50,647 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] ERROR net.schmizz.sshj.transport.TransportImpl - Dying because - Gateway authentication failed
Authentication failed
net.schmizz.sshj.transport.TransportException: Gateway authentication failed
Authentication failed
	at net.schmizz.sshj.transport.TransportImpl.gotDisconnect(TransportImpl.java:543) ~[sshj-0.35.0.jar:?]
	at net.schmizz.sshj.transport.TransportImpl.handle(TransportImpl.java:499) [sshj-0.35.0.jar:?]
	at net.schmizz.sshj.transport.Decoder.decode(Decoder.java:113) ~[sshj-0.35.0.jar:?]
	at net.schmizz.sshj.transport.Decoder.received(Decoder.java:200) ~[sshj-0.35.0.jar:?]
	at net.schmizz.sshj.transport.Reader.run(Reader.java:60) [sshj-0.35.0.jar:?]
2023-09-01 11:32:50,648 [sshj-Reader-jumpserver01.os.companyco.com/172.123.123.123:22-1693557168991] WARN  ch.cyberduck.core.sftp.SFTPSession - Disconnected BY_APPLICATION
```

---

Thank you so much for making this application, it is really a lifesaver when the only alternative is FileZilla! Really hoping this can be resolved so I can actually use it :)

**Comment by dkocher** [comment]
I can confirm that because of some fuzzy logic [^1] there is no additional prompt for the second factor requested by the server with the prompt `Gateway password: `.

[^1]: https://github.com/iterate-ch/cyberduck/blob/5e15691a52159f5b06a4cd0325d2c8376e04481c/ssh/src/main/java/ch/cyberduck/core/sftp/auth/SFTPChallengeResponseAuthentication.java#L91

**Comment by cankaratepe23** [comment]
Thanks for your help and extremely quick response. Small question: When are the snapshot builds released? I wanted to test the fix by switching to the snapshot channel and updating but doesn't seem to be working yet.

**Comment by dkocher** [comment]
> Thanks for your help and extremely quick response. Small question: When are the snapshot builds released? I wanted to test the fix by switching to the snapshot channel and updating but doesn't seem to be working yet.

The fix is included in build 8.7.0.40289. It would be great if you can confirm.

**Comment by cankaratepe23** [comment]
> The fix is included in build 8.7.0.40289. It would be great if you can confirm.

I have just tested with the build you mentioned, and can confirm that it works, although with some issues.

Everything works correctly when I connect **without** using a pre-entered or saved password; Cyberduck prompts for my password and then the one-time passcode correctly and I can connect successfully.

However, when I connect using a pre-entered password, a prompt for a "Gateway password" pops up and I have to enter my password (the one which was saved/pre-entered beforehand) manually. Only after that prompt do I get the second prompt for the one-time passcode.

Attached is the log I generated with the following steps:

1. Connect using pre-entered username/password
2. Enter the same pre-entered password into the first prompt
3. Enter the one-time passcode into the second prompt
4. Connection successful

[cyberduck.log](https://github.com/iterate-ch/cyberduck/files/12514401/cyberduck.log)

Thank you again for your help, as I can now use the app for its main functions. The issue I mentioned above is only a minor inconvenience for me.

**Comment by cankaratepe23** [comment]
A small update: I cannot reproduce the situation I described in [my previous comment to this issue](https://github.com/iterate-ch/cyberduck/issues/15076#issuecomment-1705336396) anymore.

I guess the saved password was wrong in my connection bookmark for the server.

Thanks again!
